librosa/core/constantq.py

```diff
@@ -612,115 +604,80 @@ def icqt(C, sr=22050, hop_length=512, fmin=None,
-                                           bins_per_octave=bins_per_octave,
-                                           filter_scale=filter_scale,
-                                           tuning=tuning,
-                                           window=window)[:, np.newaxis]**0.5
+    # The basis gets renormalized by the effective window length above;
+    # This step undoes that
+    fft_basis = fft_basis.todense() * n_fft / lengths[:, np.newaxis]
 
-    y = None
+    # This step conjugate-transposes the filter
+    inv_basis = np.asanyarray(fft_basis).conj().T
 
-    # Revised algorithm:
-    #   for each octave
-    #      upsample old octave
-    #      @--numba accelerate this loop?
-    #      for each basis
-    #         convolve with activation (valid-mode)
-    #         divide by window sumsquare
-    #         trim and add to total
+    # How many octaves do we have?
+    n_octaves = int(np.ceil(float(n_bins) / bins_per_octave))
 
+    y = None
     for octave in range(n_octaves - 1, -1, -1):
-        # Compute the slice index for the current octave
         slice_ = slice(-(octave+1) * bins_per_octave - 1,
                        -(octave) * bins_per_octave - 1)
 
-        # Project onto the basis
-        C_oct = C[slice_] / Cnorm[slice_]
-        basis_oct = basis[-C_oct.shape[0]:]
-
-        y_oct = None
+        # Slice this octave
+        C_oct = C[slice_]
+        inv_oct = inv_basis[:, -C_oct.shape[0]:]
 
-        # Make a dummy activation
         oct_hop = hop_length // 2**octave
-        n = n_fft + (C_oct.shape[1] - 1) * oct_hop
-
-        for i in range(basis_oct.shape[0]-1, -1, -1):
-            wss = filters.window_sumsquare(window,
-                                           n_frames,
-                                           hop_length=oct_hop,
-                                           win_length=int(lengths[i]),
-                                           n_fft=n_fft,
-                                           norm=norm)
-
-            wss *= lengths[i]**2
-
-            # Construct the response for this filter
-            y_oct_i = np.zeros(n, dtype=C_oct.dtype)
-            __activation_fill(y_oct_i, basis_oct[i], C_oct[i], oct_hop)
-            # Retain only the real part
-            # Only do window normalization for sufficiently large window
-            # coefficients
-            y_oct_i = y_oct_i.real / np.maximum(amin, wss)
-
-            if y_oct is None:
-                y_oct = y_oct_i
-            else:
-                y_oct += y_oct_i
-
-        # Remove the effects of zero-padding
-        y_oct = y_oct[n_trim:-n_trim] * bscale[i]
 
+        # Apply length corrections
+        if scale:
+            C_oct = C_oct / np.sqrt(lengths[-C_oct.shape[0]:, np.newaxis])
+        else:
+            C_oct = C_oct / (lengths[-C_oct.shape[0]:, np.newaxis] * np.sqrt(2**octave))
+
+        # Inverse-project the basis for each octave
+        D_oct = inv_oct.dot(C_oct * n_fft)
```

Comment:
inconvenient that the multiplicative factor for C_oct is broken down into multiple locations: above and here